internal/format/first_pass_visitor.go

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package format
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/emicklei/proto"
+	"github.com/uber/prototool/internal/settings"
+	"github.com/uber/prototool/internal/text"
+)
+
+type firstPassVisitor struct {
+	*baseVisitor
+
+	Syntax                *proto.Syntax
+	Package               *proto.Package
+	Options               []*proto.Option
+	ProbablyCustomOptions []*proto.Option
+	Imports               []*proto.Import
+
+	haveHitNonComment bool
+}
+
+func newFirstPassVisitor(config settings.Config) *firstPassVisitor {
+	return &firstPassVisitor{baseVisitor: newBaseVisitor(config.Format.Indent)}
+}
+
+func (v *firstPassVisitor) Do() []*text.Failure {
+	if v.Syntax != nil {
+		v.PComment(v.Syntax.Comment)
+		if v.Syntax.Comment != nil {
+			// special case
+			v.P()
```

Comment:
Usually the `syntax` line is the first line in the file. In the case there is a comment at the top of the file, I always want a newline between `syntax` and that comment. So:

```
// bla bla
syntax = "proto3";
```

Converts to:
```
// bla bla

syntax = "proto3";
```

The only time you'd generally comment above syntax is for a LICENSE header or a file-level comment. We could make it so that this doesn't format into a new line, but this was the thinking.

Reply:
Done